sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SpecifyIFormatProviderOrCultureInfo.cs

```diff
@@ -92,18 +92,99 @@ protected override void Initialize(SonarAnalysisContext context)
             ReturnsOrAcceptsFormattableType(methodSymbol) ||
             blacklistMethods.Any(x => Matches(x, methodSymbol));
 
-        private static bool HasOverloadWithFormatOrCulture(InvocationExpressionSyntax invocation, SemanticModel semanticModel) =>
-            semanticModel.GetMemberGroup(invocation.Expression)
+        private static bool HasOverloadWithFormatOrCulture(InvocationExpressionSyntax invocation, SemanticModel semanticModel)
+        {
+            return semanticModel.GetMemberGroup(invocation.Expression)
                 .OfType<IMethodSymbol>()
                 .Where(m => !m.GetAttributes(KnownType.System_ObsoleteAttribute).Any())
-                .Any(HasAnyFormatOrCultureParameter);
+                .Where(IsCompatibleOverload)
+                .Any(methodSignature => SameParametersExceptFormatOrCulture(methodSignature, GetInvocationParameters()));
+
+            IEnumerable<IParameterSymbol> GetInvocationParameters() =>
+                semanticModel.GetSymbolInfo(invocation.Expression).Symbol?.GetParameters();
+
+            // must have same number of arguments + 1 (the format or culture argument) OR is params argument
+            bool IsCompatibleOverload(IMethodSymbol m) =>
+                (m.GetParameters().Count() - invocation.ArgumentList.Arguments.Count == 1) ||
+                (m.GetParameters().Any() && m.GetParameters().Last().IsParams);
+        }
+
+        private static bool SameParametersExceptFormatOrCulture(IMethodSymbol possibleOverload, IEnumerable<IParameterSymbol> invocationParameters)
+        {
+            var parametersWithoutFormatCulture = possibleOverload.GetParameters().Where(p => !p.Type.IsAny(formatAndCultureType));

```

Comment:
This expression is evaluated multiple times in the rest of the method. Most of the following comments are about avoiding this.